### PR TITLE
Create the necessary blocks/methods for ItemRetrieval evaluation

### LINKED ISSUE #170: Top-K metrics evaluation for two-tower retrieval models

### PR DESCRIPTION
**Pull request by sararb**
Fixes #170 #109 

### Goals :soccer:
- This is a work in progress to create an optimized evaluation block for item retrieval tasks. It WILL address the following points: 

- [x]  Re-structure merin-models to separate models, tasks and blocks 
- [x] Add `pre_metrics` logic to the PredictionTask classes in order to allow for additional processing steps of predictions/targets after computing the loss and before computing metrics. 
- [x] Refactor `call_targets` method to accept a named_tuple param `PredictionOutput` containing the predictions and targets. 
- [x] Separate the functions/blocks related to item retrieval and item prediction tasks. 
- [x] Add top-k evaluation blocks to decouple top-k extraction from metrics computation.
- [x] Possibility to deactivate compute_metrics during training a retrieval model.
- [x] Define the ItemRetrieval task as a subclass of MultiClassificationTask. 
- [x] Add top-k index and implement methods that convert a trained RetrievalModel to Top-k-recommender

### Implementation Details :construction:

This PR implements the following functionalities: 

1.  **Get predictions from a trained RetrievalModel:** 
          a. Create an [IndexBlock](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/blocks/core/index.py#L31) to prepare pre-computed embeddings of items or users. 
          b. Create an [TopKIndexBlock](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/blocks/core/index.py#L122) to extract top-k candidates from indexed values. 
          c. Add the method  [to_topk_recommender](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/core.py#L2590) to RetrievalModel to get a recommender model that accepts a query embeddigns and returns the top-k scores/indices. 
          d. Add the method [_ensure_unique](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/core.py#L2511) to RetrievalModel to extract unique users or items features from a dataset of user-item interactions. 

2. **First draft of Top-k evaluation of Retrieval and  ItemPrediction tasks:** 
          a. Add [pre_metrics](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/core.py#L1860) argument to PredictionTask class that transforms predictions and targets before computing the metrics. 
          b. Add flag `compute_train_metrics` to `PredictionTask` to disable calculating heavy metrics during training. 
          c. Add [ItemRetrievalTask](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/prediction_tasks/retrieval.py#L35) class that enherits from MultiClassClassificationTask.
          d. Add the method [load_topk_evaluation](https://github.com/NVIDIA-Merlin/models/blob/e552f440e88871f9447f111e18cdb88f24b95b88/merlin/models/tf/core.py#L2596) to ItemRetrievalModel that loads items embeddings as a pre-metric block to extract top-k negatives for ranking metrics.  
         e. Add [cach_query](https://github.com/NVIDIA-Merlin/models/blob/e552f440e88871f9447f111e18cdb88f24b95b88/merlin/models/tf/blocks/retrieval/base.py#L164) flag to ItemRetrievalScorer to store query embeddings in the model context only during evaluation. 
 
3.  **Add new tf_utils:**    
        a. Add [df_to_tensor](https://github.com/NVIDIA-Merlin/models/blob/41487e977c4d749fc4765275b0187db981387187/merlin/models/tf/utils/tf_utils.py#L167) method to convert a dataframe object to a tensorflow tensor. 

### Related issues that will be addressed in separate PRs: 
- #210: Computing heavy metrics during training after each n steps.
- #208:  Optimize training of retrieval models with negative sampling.
- #203: Link top-k evaluation to batch-predict of RetrievalModel.
- #202

### CODE REVIEW

File: merlin_models/tf/core.py [hunk 1]
```diff
@@ -2217,7 +2226,7 @@ def train_step(self, inputs):
                 targets = None
 
             predictions = self(inputs, training=True)
-            loss = self.compute_loss(predictions, targets, training=True)
+            loss = self.compute_loss(predictions, targets, training=True, compute_metrics=True)
```
Comment (gabrielspmoreira) [inline merlin_models/tf/core.py:2229]: Is it always going to compute metrics during training?

File: merlin_models/tf/prediction/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    def __init__(
+        self,
+        k: int = 20,
+        **kwargs,
+    ):
+        super(BruteForceTopK, self).__init__(**kwargs)
+        self._k = k
+        self._candidates = None
+
+    def load_index(self, candidates: tf.Tensor, identifiers: tf.Tensor = None):
```
Comment (gabrielspmoreira) [inline merlin_models/tf/prediction/evaluation.py:83]: What about renaming args to `ids` and `embeddings`?

File: merlin_models/tf/utils/tf_utils.py [hunk 3]
```diff
@@ -82,8 +82,7 @@ def maybe_deserialize_keras_objects(
     return config
 
 
-def extract_topk(ks, scores, labels):
-    max_k = tf.reduce_max(ks)
+def extract_topk(max_k, scores, labels):
     topk_scores, topk_indices = tf.math.top_k(scores, max_k)
     topk_labels = gather_torch_like(labels, topk_indices, max_k)
```
Comment (gabrielspmoreira) [inline merlin_models/tf/utils/tf_utils.py:87]: Maybe later check if we could come up with a more efficient implementation of `gather_torch_like` which does not using `TensorArray`, but rather uses only `tf.gather_nd()` with a 2D index (rows, cols), like the one I implemented here [here](https://github.com/NVIDIA-Merlin/NVTabular/blob/dc9c04d126b8ad842a32a3618e8f4495afbe0678/nvtabular/loader/tensorflow.py#L404). 
Not a priority for now, but maybe it might improve the metrics performance

File: merlin_models/tf/prediction/evaluation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sararb) [reply merlin_models/tf/prediction/evaluation.py:83]: I set the names similar to what is used in TensorFlow recommenders, but how about `candidates_embeddings` and `candidates_ids`?

File: merlin_models/tf/prediction/item_prediction.py [hunk 4]
```diff
@@ -635,6 +667,13 @@ def NextItemPredictionTask(
     if normalize:
         prediction_call = L2Norm().connect(prediction_call)
 
+    pre_metrics = None
+    if len(metrics) > 0:
+        max_k = tf.reduce_max(sum([metric.top_ks for metric in metrics], []))
+        pre_metrics = ItemsPredictionTopK(
+            k=max_k, transform_to_onehot=transform_to_onehot or not sampled_softmax
```
[the remark this answers is not in the document]
Comment (sararb) [reply merlin_models/tf/prediction/item_prediction.py:674]: When `sampled_softmax == True`,  the `ItemsPredictionSampled` task is used and this task already returns a one-hot representation of targets.

File: merlin_models/tf/core.py (hunk 1 above)
Comment (gabrielspmoreira) [reply merlin_models/tf/core.py]: Or if we have a separate list of `train_metrics` and `eval_metrics`, then if `train_metrics = [] / None` that would mean that no eval metrics should be processed during training (including the `pre_metrics` call)

File: merlin_models/tf/prediction/evaluation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gabrielspmoreira) [reply merlin_models/tf/prediction/evaluation.py]: Perfect!

File: merlin/models/tf/blocks/core/index.py [hunk 5]
```diff
@@ -190,6 +190,40 @@ def call(self, inputs: tf.Tensor, k=None, **kwargs) -> Union[tf.Tensor, tf.Tenso
 
         return top_scores, top_indices
 
+    def call_targets(
```
Comment (marcromeyn) [inline merlin/models/tf/blocks/core/index.py:193]: Should we rename this to `call_outputs`?

File: merlin/models/tf/blocks/core/index.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                tf.ones([tf.shape(predictions)[0], 1]),
+                tf.zeros([tf.shape(predictions)[0], self._k]),
+            ],
+            axis=1,
+        )
+        return PredictionOutput(predictions, targets)
+
+    def compute_output_shape(self, input_shape):
+        batch_size = input_shape[0]
+        return tf.TensorShape((batch_size, self._k)), tf.TensorShape((batch_size, self._k))
```
Comment (gabrielspmoreira) [inline merlin/models/tf/blocks/core/index.py:229]: Shouldn't the tensor shapes be `(batch size, k+1)` ?

File: merlin/models/tf/core.py [hunk 7]
```diff
@@ -2504,57 +2540,106 @@ def query_embeddings(
...
+        -------
+        SequentialBlock
+        """
+        import merlin.models.tf as ml
+
+        topk_index = ml.TopKIndexBlock.from_block(
+            self.retrieval_block.item_block(), data=data, k=k - 1, context=self.context, **kwargs
+        )
+        # set cache_query to True in the ItemRetrievalScorer
+        self.loss_block.retrieval_scorer.cache_query = True
```
Comment (gabrielspmoreira) [inline merlin/models/tf/core.py:2616]: Would it make sense to have a method in `ItemRetrievalTask` (`loss_block`) to set `cache_query` and `pre_metrics`, rather than exposing those properties for other classes?

**Review by gabrielspmoreira — APPROVED** [review]
This PR ended up being large but very instrumental. I just left two small comments, but merging as it is is good to me

File: merlin/models/tf/blocks/core/index.py (hunk 6 above)
Comment (sararb) [reply merlin/models/tf/blocks/core/index.py:229]: The current implementation supposes that top-k includes the positive scores so we are retrieving `k-1` negatives.

File: merlin/models/tf/core.py (hunk 7 above)
Comment (sararb) [reply merlin/models/tf/core.py:2616]: The `loss_block` is a property of the general Model class, and I added the `retrieval_scorer` to the `ItemRetrievalTask`
